Cannot change master key

Hi there and thank you for this awesome library!

I used it to experiment with DESfire tokens and changing the master keys from DES to AES which doesn't seem to work properly in some cases.

I think that this issue is related to #4.

## Expected Behavior

Selecting the main application, authenticating and then changing the master key should work without error.

*Note:* This issues does not occur when not having selected an app before.

## Actual Behavior

When having the main app selected, there is an `ST_IntegrityError`. See the log below:

```text
2025-10-31 09:25:27,019 - desfire.DESFire - [INFO] - Selecting application with ID 00 00 00
2025-10-31 09:25:27,019 - desfire.DESFire - [DEBUG] - Preprocessing command 5A 00 00 00
2025-10-31 09:25:27,019 - desfire.DESFire - [DEBUG] - Not authenticated, skipping preprocessing
2025-10-31 09:25:27,019 - desfire.DESFire - [DEBUG] - Running APDU command, sending: 5A 00 00 00
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Received APDU response: 00 90 00
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Postprocessing PICC response 90 00
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Response is plain, returning as is
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Application selected, new authentication is needed
2025-10-31 09:25:27,030 - desfire.key - [DEBUG] - Initializing key with settings: <desfire.schemas.key_settings.KeySettings object at 0x7f403633d880>
2025-10-31 09:25:27,030 - desfire.key - [DEBUG] - Key data has been provided.
2025-10-31 09:25:27,030 - desfire.key - [DEBUG] - Setting key value to provided data
2025-10-31 09:25:27,030 - desfire.util - [DEBUG] - Converting raw data (b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00') to list of integers
2025-10-31 09:25:27,030 - desfire.util - [DEBUG] - Data is byte array. Conversion result: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,030 - desfire.key - [DEBUG] - Setting key size to 16
2025-10-31 09:25:27,030 - desfire.key - [INFO] - Initializing cipher with key type DF_KEY_AES
2025-10-31 09:25:27,030 - desfire.key - [DEBUG] - AES key type detected, setting key size to 16 bytes
2025-10-31 09:25:27,030 - desfire.key - [DEBUG] - Setting key size to 16
2025-10-31 09:25:27,030 - desfire.key - [DEBUG] - Clearing IV back to default value.
2025-10-31 09:25:27,030 - desfire.key - [DEBUG] - Setting IV to 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Authenticating against PICC
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Authenticating using AES authentication scheme and key_id 0
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Preprocessing command AA 00
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Not authenticated, skipping preprocessing
2025-10-31 09:25:27,030 - desfire.DESFire - [DEBUG] - Running APDU command, sending: AA 00
2025-10-31 09:25:27,050 - desfire.DESFire - [DEBUG] - Received APDU response: AF C3 89 C8 73 41 4F 68 31 28 94 FF 55 86 1C 3E A2
2025-10-31 09:25:27,050 - desfire.DESFire - [DEBUG] - More data present (indicated by 0xAF), returning response to callee
2025-10-31 09:25:27,050 - desfire.DESFire - [DEBUG] - Postprocessing PICC response C3 89 C8 73 41 4F 68 31 28 94 FF 55 86 1C 3E A2
2025-10-31 09:25:27,050 - desfire.DESFire - [DEBUG] - Response is plain, returning as is
2025-10-31 09:25:27,050 - desfire.DESFire - [DEBUG] - Encrypion: Random B (enc):C3 89 C8 73 41 4F 68 31 28 94 FF 55 86 1C 3E A2
2025-10-31 09:25:27,050 - desfire.key - [INFO] - Initializing cipher with key type DF_KEY_AES
2025-10-31 09:25:27,050 - desfire.key - [DEBUG] - AES key type detected, setting key size to 16 bytes
2025-10-31 09:25:27,050 - desfire.key - [DEBUG] - Setting key size to 16
2025-10-31 09:25:27,050 - desfire.key - [DEBUG] - Clearing IV back to default value.
2025-10-31 09:25:27,050 - desfire.key - [DEBUG] - Setting IV to 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,050 - desfire.util - [DEBUG] - Creating cipher module for key type DF_KEY_AES
2025-10-31 09:25:27,050 - desfire.util - [DEBUG] - Creating AES cipher module
2025-10-31 09:25:27,051 - desfire.key - [DEBUG] - Decrypting data: C3 89 C8 73 41 4F 68 31 28 94 FF 55 86 1C 3E A2 using key type DF_KEY_AES
2025-10-31 09:25:27,051 - desfire.DESFire - [DEBUG] - Encrypion: Random B (dec): 44 58 CD 81 9D 26 61 E7 A1 28 B9 7D AA 58 AB 59
2025-10-31 09:25:27,051 - desfire.DESFire - [DEBUG] - Encrypion: Random B (dec, rot): 58 CD 81 9D 26 61 E7 A1 28 B9 7D AA 58 AB 59 44
2025-10-31 09:25:27,051 - desfire.util - [DEBUG] - Converting raw data (b'\x99l\x1d\xcc\xa8\xc3\xb4\x03\x1a\x12\xb1\xce\x1bx\x95\x13') to list of integers
2025-10-31 09:25:27,051 - desfire.util - [DEBUG] - Data is byte array. Conversion result: 99 6C 1D CC A8 C3 B4 03 1A 12 B1 CE 1B 78 95 13
2025-10-31 09:25:27,051 - desfire.DESFire - [DEBUG] - Encrypion: Random A: 99 6C 1D CC A8 C3 B4 03 1A 12 B1 CE 1B 78 95 13
2025-10-31 09:25:27,051 - desfire.DESFire - [DEBUG] - Encrypion: Random AB: 99 6C 1D CC A8 C3 B4 03 1A 12 B1 CE 1B 78 95 13 58 CD 81 9D 26 61 E7 A1 28 B9 7D AA 58 AB 59 44
2025-10-31 09:25:27,051 - desfire.key - [DEBUG] - Setting IV to C3 89 C8 73 41 4F 68 31 28 94 FF 55 86 1C 3E A2
2025-10-31 09:25:27,051 - desfire.util - [DEBUG] - Creating cipher module for key type DF_KEY_AES
2025-10-31 09:25:27,051 - desfire.util - [DEBUG] - Creating AES cipher module
2025-10-31 09:25:27,051 - desfire.DESFire - [DEBUG] - Encrypion: Random AB (enc): 61 EB 4B 2E 69 82 4D 1C D7 2F 87 5E BF 58 CF DA 54 A3 0A E9 2B 15 05 4F 4B C8 41 07 38 19 81 E2
2025-10-31 09:25:27,051 - desfire.DESFire - [DEBUG] - Preprocessing command AF 61 EB 4B 2E 69 82 4D 1C D7 2F 87 5E BF 58 CF DA 54 A3 0A E9 2B 15 05 4F 4B C8 41 07 38 19 81 E2
2025-10-31 09:25:27,051 - desfire.DESFire - [DEBUG] - Not authenticated, skipping preprocessing
2025-10-31 09:25:27,052 - desfire.DESFire - [DEBUG] - Running APDU command, sending: AF 61 EB 4B 2E 69 82 4D 1C D7 2F 87 5E BF 58 CF DA 54 A3 0A E9 2B 15 05 4F 4B C8 41 07 38 19 81 E2
2025-10-31 09:25:27,085 - desfire.DESFire - [DEBUG] - Received APDU response: 00 D7 BF E8 6E E0 22 20 1C 54 82 B0 93 3F F0 8A 0C
2025-10-31 09:25:27,085 - desfire.DESFire - [DEBUG] - Postprocessing PICC response D7 BF E8 6E E0 22 20 1C 54 82 B0 93 3F F0 8A 0C
2025-10-31 09:25:27,085 - desfire.DESFire - [DEBUG] - Response is plain, returning as is
2025-10-31 09:25:27,085 - desfire.DESFire - [DEBUG] - Encrypion: Random A (enc): D7 BF E8 6E E0 22 20 1C 54 82 B0 93 3F F0 8A 0C
2025-10-31 09:25:27,085 - desfire.key - [DEBUG] - Setting IV to 54 A3 0A E9 2B 15 05 4F 4B C8 41 07 38 19 81 E2
2025-10-31 09:25:27,085 - desfire.util - [DEBUG] - Creating cipher module for key type DF_KEY_AES
2025-10-31 09:25:27,085 - desfire.util - [DEBUG] - Creating AES cipher module
2025-10-31 09:25:27,085 - desfire.key - [DEBUG] - Decrypting data: D7 BF E8 6E E0 22 20 1C 54 82 B0 93 3F F0 8A 0C using key type DF_KEY_AES
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Encrypion: Random A (dec): 6C 1D CC A8 C3 B4 03 1A 12 B1 CE 1B 78 95 13 99
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Encrypion: Random A (dec, rot): 99 6C 1D CC A8 C3 B4 03 1A 12 B1 CE 1B 78 95 13
2025-10-31 09:25:27,086 - desfire.DESFire - [INFO] - Authentication successful
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Encrypion: Calculating Session key
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting key value to provided data
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Converting raw data (b'\x99l\x1d\xccDX\xcd\x81\x1bx\x95\x13\xaaX\xabY') to list of integers
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Data is byte array. Conversion result: 99 6C 1D CC 44 58 CD 81 1B 78 95 13 AA 58 AB 59
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting key size to 16
2025-10-31 09:25:27,086 - desfire.cmac - [DEBUG] - Initializing CMAC with provided key. Calculating K1 and K2.
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Creating cipher module for key type DF_KEY_AES
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Creating AES cipher module
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Clearing IV back to default value.
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting IV to 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Initializing key with settings: <desfire.schemas.key_settings.KeySettings object at 0x7f403633e1e0>
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Key data has been provided.
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting key value to provided data
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Converting raw data (b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00') to list of integers
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Data is byte array. Conversion result: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting key size to 16
2025-10-31 09:25:27,086 - desfire.key - [INFO] - Initializing cipher with key type DF_KEY_AES
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - AES key type detected, setting key size to 16 bytes
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting key size to 16
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Clearing IV back to default value.
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting IV to 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,086 - applicator - [DEBUG] - Authenticated to main app with aes. Changing key to default AES null key.
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Initializing key with settings: <desfire.schemas.key_settings.KeySettings object at 0x7f403633e1e0>
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Key data has been provided.
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting key value to provided data
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Converting raw data (b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00') to list of integers
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Data is byte array. Conversion result: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting key size to 16
2025-10-31 09:25:27,086 - desfire.key - [INFO] - Initializing cipher with key type DF_KEY_AES
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - AES key type detected, setting key size to 16 bytes
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting key size to 16
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Clearing IV back to default value.
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting IV to 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,086 - desfire.DESFire - [INFO] - Executing command: change_key (0xc4) for key 0
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Changing the key we're authenticated with, need to re-authenticate after
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Calculating CRC32 checksum for data: C4 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Converting raw data (461770506) to list of integers
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Data is integer. Conversion result: 0A 0F 86 1B
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Checksum: 0A 0F 86 1B
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Preprocessing command C4 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0A 0F 86 1B
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Command requires data to be encrypted. Calculating CRC and encrypting message
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Original data: C4 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0A 0F 86 1B
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Creating cipher module for key type DF_KEY_AES
2025-10-31 09:25:27,086 - desfire.util - [DEBUG] - Creating AES cipher module
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Encrypted data: C4 00 CB F8 00 FE 67 FF 85 CA 1E B2 BE F6 24 6D B0 B0 AA 1D 7A DF E9 7D E5 86 AD 8E 53 52 82 26 DF 88
2025-10-31 09:25:27,086 - desfire.key - [DEBUG] - Setting IV to AA 1D 7A DF E9 7D E5 86 AD 8E 53 52 82 26 DF 88
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Running APDU command, sending: C4 00 CB F8 00 FE 67 FF 85 CA 1E B2 BE F6 24 6D B0 B0 AA 1D 7A DF E9 7D E5 86 AD 8E 53 52 82 26 DF 88
2025-10-31 09:25:27,112 - desfire.DESFire - [DEBUG] - Received APDU response: 1E 90 00
2025-10-31 09:25:27,112 - desfire.DESFire - [ERROR] - Received error from card: ST_IntegrityError
```

## Solution

Especially interesting is this line:

```text
2025-10-31 09:25:27,086 - desfire.DESFire - [DEBUG] - Preprocessing command C4 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0A 0F 86 1B
```

This shows that despite me selecting AES as key type, the second byte does not have the 7th bit (`0x80`) set. Analyzing the source code, I also do not find the [log message](https://github.com/waza-ari/python-desfire/blob/806d802d9de49d332a03f15b4fafe50e85afa36d/src/desfire/DESFire.py#L707) that tells us about the key number parameter being calculated.

I found that the **comparison whether the main application is selected does not work as expected.**, as can be seen in the following snippet:

https://github.com/waza-ari/python-desfire/blob/806d802d9de49d332a03f15b4fafe50e85afa36d/src/desfire/DESFire.py#L700-L707

The issue is that `last_selected_application = [0x00, 0x00, 0x00]` when the main application was selected and, thus, the comparison fails.

If you agree with this and are fine with it, I can provide a fix.

--- a/src/desfire/DESFire.py
+++ b/src/desfire/DESFire.py
@@ -561,7 +561,7 @@ class DESFire:
             DESFireCommunicationMode.CMAC if self.is_authenticated else DESFireCommunicationMode.PLAIN,
         )
         res = KeySettings(
-            application_id=self.last_selected_application or [0x0],
+            application_id=self.last_selected_application or [0x00, 0x00, 0x00],
             key_type=DESFireKeyType(resp[1] & 0xF0),  # Only interested in first 4 bits of the second byte
             max_keys=resp[1] & 0x0F,  # Only interested in last 4 bits of the second byte
             settings=[],
@@ -697,7 +697,10 @@ class DESFire:
         # The type of key can only be changed for the PICC master key
         # Applications must define their key type in create_application()
         key_number = key_id & 0x0F
-        if self.last_selected_application == [0x00] or self.last_selected_application is None:
+        # Changing the master key requires special flags being set.
+        # Therefore, we need to detect whether we're trying to change the PICC master key on application 0x000000
+        # See https://github.com/waza-ari/python-desfire/issues/20 and https://github.com/waza-ari/python-desfire/issues/4
+        if self.last_selected_application is None or self.last_selected_application == [0x00, 0x00, 0x00]:
             key_number = key_number | current_key.key_type.value
             # changing the PICC master key type requires special key_numbers
             if new_key.key_type == DESFireKeyType.DF_KEY_AES:

